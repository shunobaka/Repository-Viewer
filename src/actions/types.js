export const USER_LOADED = 'USER_LOADED';
export const USERS_SEARCH_RESULT = 'USERS_SEARCH_RESULT';
export const REPOSITORIES_LOADED = 'REPOSITORIES_LOADED';
export const REPOSITORIES_FILTERED = 'REPOSITORIES_FILTERED';
export const ADD_ALERT = 'ADD_ALERT';
