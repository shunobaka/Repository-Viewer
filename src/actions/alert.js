/**
 * @fileoverview Defines functions that return action objects used by redux to
 *    update alert state.
 */
import { ADD_ALERT, REMOVE_ALERT, CLEAR_ALERTS } from './types';

/**
 * Returns a redux action object based on the provided alert object.
 * @param {object} alert An object representing an alert with fields message and id
 * @return {object} Redux action object
 * @export
 */
export const addAlert = (alert) => {
  return {
    type: ADD_ALERT,
    payload: alert,
  };
};

/**
 * Returns a redux action object used to remove the alert with the provided id.
 * @param {string} id The uuid of the alert that is to be removed
 * @return {object} Redux action object
 * @export
 */
export const removeAlert = (id) => {
  return {
    type: REMOVE_ALERT,
    payload: id,
  };
};

/**
 * Returns a redux action object used to remove all alerts.
 * @return {object} Redux action object
 * @export
 */
export const clearAlerts = () => {
  return {
    type: CLEAR_ALERTS,
  };
};
