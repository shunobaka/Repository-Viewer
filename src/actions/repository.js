/**
 * @fileoverview Defines functions that return action objects used by redux to
 *    update repository state.
 */
import { REPOSITORIES_LOADED, REPOSITORIES_FILTERED } from './types';

/**
 * Returns a redux action object used to save repositories in state based on the
 *    provided repositories.
 * @param {array} repos Repository objects to be saved in redux state
 * @return {object} Redux action object
 * @export
 */
export const repositoriesLoaded = (repos) => {
  return {
    type: REPOSITORIES_LOADED,
    payload: repos,
  };
};

/**
 * Returns a redux action object used to filter repositories in state based on the
 *    provided query.
 * @param {string} query The name query by which repositories are filtered
 * @return {object} Redux action object
 * @export
 */
export const repositoriesFiltered = (query) => {
  return {
    type: REPOSITORIES_FILTERED,
    payload: query,
  };
};
