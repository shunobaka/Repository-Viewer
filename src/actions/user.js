/**
 * @fileoverview Defines functions that return action objects used by redux to
 *    update user state.
 */
import { USERS_SEARCH_RESULT, USER_LOADED } from './types';

/**
 * Returns a redux action object used to save users and username query in state
 *    based on the provided users and query.
 * @param {array} users User objects to be saved in the redux state
 * @param {*} query The username query by which the users were retrieved
 * @returns Redux action object
 * @export
 */
export const usersLoaded = (users, query) => {
  return {
    type: USERS_SEARCH_RESULT,
    payload: {
      users,
      query,
    },
  };
};

/**
 * Returns a redux action object used to save user in state based on the provided
 *    user.
 * @param {object} user User object to be saved in redux state
 * @returns Redux action object
 * @export
 */
export const userLoaded = (user) => {
  return {
    type: USER_LOADED,
    payload: user,
  };
};
