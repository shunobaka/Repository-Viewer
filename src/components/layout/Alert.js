/**
 * @fileoverview Defines an Alert react component that uses the alert redux state
 *    to display alerts.
 */
import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Container } from 'react-bootstrap';

/**
 * Alert react component that displays provided alerts. Gets dynamically updated when
 *    the alert redux state changes.
 * @param {object} props Contains the mapped alerts from the alert redux state
 */
const Alert = ({ alerts }) => {
  return (
    <Fragment>
      {alerts.length > 0 && (
        <div className="alert-container text-center">
          <Container>
            {alerts.map((alert, index) => {
              return (
                <div className="alert alert-danger" key={alert.id} role="alert">
                  {alert.message}
                </div>
              );
            })}
          </Container>
        </div>
      )}
    </Fragment>
  );
};

Alert.propTypes = {
  alerts: PropTypes.array.isRequired,
};

/**
 * Maps the alert redux state to the props consumed by the react component.
 * @param {object} state The complete redux state object
 */
const mapStateToProps = (state) => ({
  alerts: state.alert,
});

export default connect(mapStateToProps)(Alert);
