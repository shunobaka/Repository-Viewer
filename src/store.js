/**
 * @fileoverview Creates and exports the redux store.
 */
import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';

import combinedReducer from './reducers';

const middleware = [thunk];

/**
 * The redux store that is created and exported.
 * @export
 */
const store = createStore(
  combinedReducer,
  composeWithDevTools(applyMiddleware(...middleware))
);

export default store;
