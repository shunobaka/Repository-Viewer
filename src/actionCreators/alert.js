import { addAlert, removeAlert } from '../actions/alert';
import uuid from '../utils/uuid';

export const setAlert = (message, remove = true, timeout = 5000) => async (
  dispatch
) => {
  const id = uuid();
  const alert = {
    message,
    id,
  };

  dispatch(addAlert(alert));

  if (remove) {
    setTimeout(() => dispatch(removeAlert(id)), timeout);
  }
};

export default {
  setAlert,
};
