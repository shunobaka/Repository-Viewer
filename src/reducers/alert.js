import { ADD_ALERT } from '../actions/types';

const initialState = [];

export default function (state = initialState, action) {
  const { type, payload } = action;

  switch (type) {
    case ADD_ALERT:
      return [...state, payload];
    default:
      return state;
  }
}
