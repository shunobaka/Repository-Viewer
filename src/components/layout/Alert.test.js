import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { Provider } from 'react-redux';
import Alert from './Alert';
import store from '../../store';
import { addAlert } from '../../actions/alert';

describe('alert snapshots', () => {
  const mockAlert = {
    message: 'Test',
    id: 0,
  };

  it('no alerts renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Alert />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('single alert renders correctly', () => {
    act(() => {
      store.dispatch(addAlert(mockAlert));
    });

    const tree = renderer
      .create(
        <Provider store={store}>
          <Alert />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
