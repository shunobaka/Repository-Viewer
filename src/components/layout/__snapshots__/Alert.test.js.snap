// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`alert snapshots no alerts renders correctly 1`] = `null`;

exports[`alert snapshots single alert renders correctly 1`] = `
<div
  className="alert-container text-center"
>
  <div
    className="container"
  >
    <div
      className="alert alert-danger"
      role="alert"
    >
      Test
    </div>
  </div>
</div>
`;
